### ISSUE TITLE
asynctest.mock.patch() doesn't patch correctly generators (and undecorated coroutines)

### ISSUE THREAD

**Issue by sjakthol** [Martiusweb/asynctest#5]
### Setup
You have an external package that has an async method your modules call and you want to mock out the calls to the external async method in your unittests.

### The problem
When you decorate your your test methods with `@asynctest.mock.patch`, the external method is not correctly patched and the original method is called. However, if you patch the external method using a `with` statement inside the test case, the method is patched correctly and the mock is called instead of the original one.

Testing a similar scenario with sync methods and the unittest package, `unittest.mock.patch` works properly both as a decorator and as a context manager.

I've setup a small demo about this issue in https://github.com/sjakthol/asynctest-patch-issue-demo to ease reproducing this bug.

**Comment by Martiusweb** [comment]
Hi Sami,

This is the very first bug report for the asynctest project! Thank you for reporting with a comprehensive test case.

__tl;dr:__ The bug appears because your test function (``test_async.AsyncTestCase.test_decorator``) is not decorated with ``@asyncio.coroutine``.

In fact, the patcher will only patch the target (``my_package.gen.generate``) during the instantiation of the coroutine:

    @asynctest.mock.patch("async_package.gen.generate", return_value="mock")
    def test_decorator(self, mock):
        self.assertEqual((yield from my_package.module.method_async()), "mock")

is equivalent to:

    def test_ctx_manager(self):
        with asynctest.mock.patch("async_package.gen.generate",
                                  return_value="mock"):
            coro = my_package.module.method_async()
        self.assertEqual((yield from coro), "mock")  # not patched anymore

This is wrong because when the coroutine is awaited (yielded from), the original method is already restored by the patcher.

The test is done in ``asynctest.mock._patch.decorate_callable()`` ([mock.py:346](https://github.com/Martiusweb/asynctest/blob/0cabeaa/asynctest/mock.py#L346)).

I'm not sure yet if I should address the issue (treat generator functions as coroutine functions even if the decorator is missing) since it may not be strictly future proof. On the other hand, currently the order or the decorators matters, as:

    @patch(...)
    @asyncio.coroutine
    def test_case(self, mock):
       # is okay

while

    @asyncio.coroutine
    @patch(...)
    def test_case(self, mock):
       # is not: the patched object is not a coroutine when passed to the decorator

**Comment by sjakthol** [comment]
I wonder why I did not think of marking my async test cases with the coroutine decorator. It would've been the obvious thing to do given that the tests were async... Well, you'll always learn something new.

But anyway, thank you for the great explanation and keep up the good work!

**Comment by Martiusweb** [comment]
I'll use this issue to keep my thoughts. Feel free to unsubscribe if you don't want to be bothered by the notifications :)

Note that ``unittest.mock.patch`` doesn't patch generators.

The way ``asynctest`` currently patch coroutines may not be correct, since the patcher is effective as long a the coroutine did not terminate, which can lead to surprising behaviors.

Considering this example:

    @asynctest.mock.patch('sync_package.gen.generate', return_value='mock')
    @asyncio.coroutine
    def patched(mock):
        while True:
            yield my_package.module.method_sync()

    gen = iter(patched())

    print("expect mock:", next(gen))
    print("mock or original?", my_package.module.method_sync())
    print("expect mock:", next(gen))
    del gen
    print("mock or original?", my_package.module.method_sync())

While we can expect that inside the coroutine, the target is still patched, I'd say that it should not be the case when called outside of its scope.

Instead of wrapping the coroutine function, only ``__next__``, and ``__await__`` should be patched.

**Comment by ZhenTek** [comment]
using context manger instead of decorator,  with patch.object(ProductionClass, 'method', return_value=None) as mock_method:
https://docs.python.org/3/library/unittest.mock.html#patch-object
OR
add mock_<mocked_method_name> to test method parameters
